### PR TITLE
dynamic ACCESS-CONTROL-ALLOW-ORIGIN header

### PR DESCRIPTION
**Pull request by nmonterroso**
@Wikia/cake @vforge 

As a followup to Wikia/app#13477 I'm updating `CrossOriginResourceSharingHeaderHelper` to dynamically set the `Access-Control-Allow-Origin` header depending on the value of the `Origin` header in the request. It needs to be this way to support multiple values.

Depending on whether or not the request is in dev/prod the default value of allowed origins will be `[.wikia-dev.us, .wikia-dev.pl]` or `[.wikia.com]`. This allows us to develop from any dev domain and have CORS work as we'd expect them to. Additionally I've added a function `setAllowAllOrigins()` that responds with `*` for the value of the header.

### CODE REVIEW

File: includes/wikia/helpers/CrossOriginResourceSharingHeaderHelper.php [hunk 1]
```diff
@@ -88,14 +106,26 @@ private function setResponseHeader( WikiaResponse $response, $headerName, $value
...
 
-	/**
-	 * @return $this
-	 */
-	public function readConfig() {
-		global $wgCORSAllowOrigin;
+	public function allowWhitelistedOrigins(Array $additionalOrigins = [] ) {
+		global $wgWikiaEnvironment, $wgCORSAllowOrigin;
+
+		switch ( $wgWikiaEnvironment ) {
```
Comment (WikiaTech) [inline includes/wikia/helpers/CrossOriginResourceSharingHeaderHelper.php]: MINOR Replace this "switch" statement with "if" statements to increase readability. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1301)

**Comment by WikiaTech** [comment]
SonarQube analysis reported 2 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [ArticlesApiController.class.php#L902](https://github.com/Wikia/app/blob/0b465fa5cb216ebfe7cacc4db3a9923bff2a8ada/includes/wikia/api/ArticlesApiController.class.php#L902): Remove the unused function parameter "$caller". [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1172)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by WikiaTech** [comment]
SonarQube analysis reported 2 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [ArticlesApiController.class.php#L902](https://github.com/Wikia/app/blob/057fe54a11b21f592e392dbcb94e81ebaf628d18/includes/wikia/api/ArticlesApiController.class.php#L902): Remove the unused function parameter "$caller". [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1172)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: includes/wikia/helpers/CrossOriginResourceSharingHeaderHelper.php [hunk 2]
```diff
@@ -69,9 +71,26 @@ public function setHeaders( WikiaResponse $response, $mergeExisting = true ) {
...
 
+		$this->setOriginHeader($response);
+
 		return $this;
 	}
 
+	private function setOriginHeader( WikiaResponse $response ) {
+		if ( $this->allowAllOrigins ) {
+			$this->setResponseHeader( $response, self::ALLOW_ORIGIN_HEADER_NAME, '*' );
+		} else if ( count( $this->whitelistOrigins ) > 0 && isset($_SERVER['HTTP_ORIGIN']) ) {
```
Comment (lizlux) [inline includes/wikia/helpers/CrossOriginResourceSharingHeaderHelper.php]: parentheses whitespace is needed here and below.

**Comment by drsnyder** [comment]
LGTM

File: includes/wikia/helpers/CrossOriginResourceSharingHeaderHelper.php (hunk 2 above)
Comment (nmonterroso) [reply includes/wikia/helpers/CrossOriginResourceSharingHeaderHelper.php]: 5431cd1

File: includes/wikia/helpers/CrossOriginResourceSharingHeaderHelper.php [hunk 3]
```diff
@@ -88,14 +107,26 @@ private function setResponseHeader( WikiaResponse $response, $headerName, $value
...
 
-	/**
-	 * @return $this
-	 */
-	public function readConfig() {
-		global $wgCORSAllowOrigin;
+	public function allowWhitelistedOrigins( Array $additionalOrigins = [] ) {
+		global $wgWikiaEnvironment, $wgCORSAllowOrigin;
+
+		switch ( $wgWikiaEnvironment ) {
```
Comment (WikiaTech) [inline includes/wikia/helpers/CrossOriginResourceSharingHeaderHelper.php:113]: MINOR Replace this "switch" statement with "if" statements to increase readability. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1301)

**Comment by WikiaTech** [comment]
SonarQube analysis reported 2 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [ArticlesApiController.class.php#L902](https://github.com/Wikia/app/blob/5431cd1b5b76f7d28568f706742bc32cce1ed137/includes/wikia/api/ArticlesApiController.class.php#L902): Remove the unused function parameter "$caller". [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1172)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**